File: packages/core/src/research_engine/domain/extractions.py
"""Extraction framework domain types."""

from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from research_engine.domain.common import ExtractionStatus


class ExtractionSchema(BaseModel):
    """A registered extraction schema."""

    id: UUID
    name: str
    version: int
    owner: str
    schema_def: dict[str, Any] = Field(alias="schema")
    prompt_template: str
    created_at: datetime

    model_config = {"populate_by_name": True}


class ExtractionSchemaDraft(BaseModel):
    """Data needed to register an extraction schema."""

    name: str
    version: int
    owner: str
    schema_def: dict[str, Any]
    prompt_template: str


class Extraction(BaseModel):
    """One row per (passage x schema x extractor_version) invocation."""

    id: UUID
    passage_id: UUID
    schema_id: UUID
    extractor_version: str
    llm_model: str
    status: ExtractionStatus
    error: str | None = None
    records: list[dict[str, Any]] = Field(default_factory=list)
    llm_call_id: UUID | None = None
    created_at: datetime


class ExtractionRecord(BaseModel):
    """A single extracted record, materialized for queryability."""

    id: UUID
    extraction_id: UUID
    passage_id: UUID
    schema_id: UUID
    record_type: str
    data: dict[str, Any]
    evidence_start: int | None = None
    evidence_end: int | None = None
    created_at: datetime


class ExtractionOptions(BaseModel):
    """Options for running an extraction."""

    force_refresh: bool = False
    llm_model: str | None = None
    concurrency: int = 8
    batch_size: int = 10
    retry_on_validation_error: bool = True
    caller: str = "core"


class ExtractionResult(BaseModel):
    """Result of extracting from a single passage."""

    passage_id: UUID
    status: ExtractionStatus
    records: list[dict[str, Any]] = Field(default_factory=list)
    from_cache: bool = False
    llm_call_id: UUID | None = None
    error: str | None = None

    @classmethod
    def from_cached(cls, extraction: Extraction) -> ExtractionResult:
        return cls(
            passage_id=extraction.passage_id,
            status=extraction.status,
            records=extraction.records,
            from_cache=True,
            llm_call_id=extraction.llm_call_id,
        )


class ExtractionBatch(BaseModel):
    """Result of a batch extraction."""

    results: list[ExtractionResult]
    schema_name: str
    schema_version: int

File: crates/marginalia-types/src/extractions.rs
//! Extraction framework types mirroring `domain/extractions.py`.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use uuid::Uuid;

use crate::common::ExtractionStatus;

/// A registered extraction schema.
///
/// `schema_def` serializes as `schema` on the wire (`alias="schema"`,
/// `populate_by_name` in Python).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExtractionSchema {
    pub id: Uuid,
    pub name: String,
    pub version: i64,
    pub owner: String,
    /// Accepts `schema_def` too: Python's default dump uses the field name.
    #[serde(rename = "schema", alias = "schema_def")]
    pub schema_def: Map<String, Value>,
    pub prompt_template: String,
    pub created_at: DateTime<Utc>,
}

/// Data needed to register an extraction schema.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExtractionSchemaDraft {
    pub name: String,
    pub version: i64,
    pub owner: String,
    pub schema_def: Map<String, Value>,
    pub prompt_template: String,
}

/// One row per (passage x schema x extractor_version) invocation.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Extraction {
    pub id: Uuid,
    pub passage_id: Uuid,
    pub schema_id: Uuid,
    pub extractor_version: String,
    pub llm_model: String,
    pub status: ExtractionStatus,
    #[serde(default)]
    pub error: Option<String>,
    #[serde(default)]
    pub records: Vec<Map<String, Value>>,
    #[serde(default)]
    pub llm_call_id: Option<Uuid>,
    pub created_at: DateTime<Utc>,
}

/// A single extracted record, materialized for queryability.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExtractionRecord {
    pub id: Uuid,
    pub extraction_id: Uuid,
    pub passage_id: Uuid,
    pub schema_id: Uuid,
    pub record_type: String,
    pub data: Map<String, Value>,
    #[serde(default)]
    pub evidence_start: Option<i64>,
    #[serde(default)]
    pub evidence_end: Option<i64>,
    pub created_at: DateTime<Utc>,
}

/// Options for running an extraction.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExtractionOptions {
    #[serde(default)]
    pub force_refresh: bool,
    #[serde(default)]
    pub llm_model: Option<String>,
    #[serde(default = "default_concurrency")]
    pub concurrency: i64,
    #[serde(default = "default_batch_size")]
    pub batch_size: i64,
    #[serde(default = "default_true")]
    pub retry_on_validation_error: bool,
    #[serde(default = "default_caller")]
    pub caller: String,
}

fn default_concurrency() -> i64 {
    8
}
fn default_batch_size() -> i64 {
    10
}
fn default_true() -> bool {
    true
}
fn default_caller() -> String {
    "core".to_owned()
}

/// Result of extracting from a single passage.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExtractionResult {
    pub passage_id: Uuid,
    pub status: ExtractionStatus,
    #[serde(default)]
    pub records: Vec<Map<String, Value>>,
    #[serde(default)]
    pub from_cache: bool,
    #[serde(default)]
    pub llm_call_id: Option<Uuid>,
    #[serde(default)]
    pub error: Option<String>,
}

impl ExtractionResult {
    /// The `from_cached` constructor.
    pub fn from_cached(extraction: &Extraction) -> Self {
        Self {
            passage_id: extraction.passage_id,
            status: extraction.status,
            records: extraction.records.clone(),
            from_cache: true,
            llm_call_id: extraction.llm_call_id,
            error: None,
        }
    }
}

/// Result of a batch extraction.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExtractionBatch {
    pub results: Vec<ExtractionResult>,
    pub schema_name: String,
    pub schema_version: i64,
}
